### PR TITLE
Replacement docker role

### PR DESCRIPTION
**Pull request by manics**
This is a new all-in-one docker role, which replaces `docker-storage` and `docker-advanced`.

I've also renamed the `omero-basedeps` role to `basedeps`.

Example basic playbook:
```
- hosts: docker
  roles:
  - role: docker
    # Optional:
    docker_use_upstream_repo: True
```
This will install the upstream docker with the default configuration (uses loopback devices and the default network)

Example advanced playbook:
```
- hosts: docker-advanced
  roles:
  - role: basedeps
  - role: upgrade-distpackages
  - role: docker
    docker_use_upstream_repo: True
    # Custom LVM storage partitions
    docker_use_custom_storage: True
    docker_vgname: VolGroup00
    docker_poolsize: 20G
    docker_metadatasize: 200M
    docker_basefs: xfs
    docker_volumesize: 10G
    docker_lvfilesystem: xfs
    docker_groupmembers: userx
    # Use the standard docker bridge but use a subset of the IPs for testing
    docker_use_custom_network: True
    docker_bridge_name: docker0
    docker_bridge_ips: 172.17.123.0/24
    # Try a local docker DNS setup
  - role: docker-dns-server
    docker_dns:
      forwarders: 8.8.8.8:53,8.8.4.4:53
      domain: internal
      etcd: etcd1
  - role: docker-dns-client
    docker_dns:
      domain: internal
      etcd: etcd1
```
This installs upstream docker, configures the docker partitions (for the docker containers, and also `/var/lib/docker`), modifies the network config slightly so only part of the default IP range is used. It also installs a local docker DNS server.

Testing:
- Check `docker info` matches the configured values
- Check the IP assigned to a container `docker run ...` is within the configured IP range
- This should return the (internal) IPs for running containers `dig \*.docker.internal @test.hostname  SRV`

## Original description PR (for the record):

I've renamed the old `docker` roles to `docker-advanced` (which requires `docker-storage`), and added a default `docker` role (requires storage sizes to be set, use default docker networking). Also prefixed all(?) docker vars.

In theory this should run against the existing servers without making changes, in practice when run in check mode it suggests it may... I'm still investigating.

It should be possible to test the `docker` role now on a VM or the new docker server, see roles/docker/README.md

### CODE REVIEW

**Comment by joshmoore** [comment]
And just to clarify, you'd now expect the default `docker` role to be used on production/dogfish/minke/thing?

**Comment by manics** [comment]
Yes, I'll test it this week

**Comment by manics** [comment]
I think this can be tested now.

File: ansible/roles/docker-advanced/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Role Variables
+--------------
+
+These variables must be defined (defaults aren't provided):
+
+- `docker_vgname`: The volume group used in `docker-storage`. This is necessary because the docker setup process requires the name of the raw logical volume device.
+
+Optional variables:
+
+- `docker_bridge.name`: The name of a custom network bridge for docker
```
Comment (mtbc) [inline ansible/roles/docker-advanced/README.md]: these don't end with periods but those in the other sections do

**Comment by joshmoore** [comment]
So test via https://github.com/manics/infrastructure/blob/docker-auto/ansible/roles/docker-advanced/README.md? Suggestions on where?

**Comment by manics** [comment]
You mean https://github.com/manics/infrastructure/tree/docker-auto/ansible/roles/docker ?
infra-testpr?

**Comment by joshmoore** [comment]
Steps for testing:

```
$ mkdir c7
$ cd c7
$ vagrant init centos/7
$ vagrant up --provider virtualbox # After all kinds of brew upgrades...
$ vagrant ssh

# In the box
$  sudo yum install git
$  git clone -b docker-auto git://github.com/manics/infrastructure
$  git clone git://github.com/ansible/ansible
$  cd ansible/
$  git submodule update --init --recursive
$  export PATH=$PATH:`pwd`/bin
$  export PYTHONPATH=`pwd`/lib
$  export ANSIBLE_ROLES_PATH=`pwd`/infrastructure/ansible/roles
$  sudo yum install PyYAML
$  sudo yum install python-jinja2
$  vi host
$  vi play.yml
$  ansible-playbook -i host play.yml -b

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [lvm-partition : storage | create logical volume] *************************
skipping: [localhost]

TASK [lvm-partition : storage | format] ****************************************
skipping: [localhost]

TASK [lvm-partition : storage | mount] *****************************************
skipping: [localhost]

TASK [docker : include] ********************************************************
skipping: [localhost]

TASK [docker : include] ********************************************************
included: /home/vagrant/infrastructure/ansible/roles/docker/tasks/docker-upstream.yml for localhost

TASK [docker : upstream docker | setup repository] *****************************
changed: [localhost]

TASK [docker : upstream docker | install docker] *******************************
changed: [localhost]

TASK [docker : include] ********************************************************
included: /home/vagrant/infrastructure/ansible/roles/docker/tasks/advanced-common.yml for localhost

TASK [docker : storage | setup lvm docker-pool] ********************************
skipping: [localhost]

TASK [docker : include] ********************************************************
skipping: [localhost]

TASK [docker : include] ********************************************************
included: /home/vagrant/infrastructure/ansible/roles/docker/tasks/advanced-upstream.yml for localhost

TASK [docker : upstream docker | configure advanced] ***************************
skipping: [localhost]

TASK [docker : Check change-time of original docker.service] *******************
ok: [localhost]

TASK [docker : Check change-time of modified docker.service] *******************
ok: [localhost]

TASK [docker : upstream docker | copy systemd file] ****************************
skipping: [localhost]

TASK [docker : upstream docker | customise systemd file] ***********************
skipping: [localhost]

TASK [docker : upstream docker | remove custom systemd file] *******************
ok: [localhost]

TASK [docker : include] ********************************************************
included: /home/vagrant/infrastructure/ansible/roles/docker/tasks/docker-common.yml for localhost

TASK [docker : docker | enable] ************************************************
changed: [localhost]

TASK [docker : docker | group members] *****************************************
skipping: [localhost] => (item=docker_groupmembers)

PLAY RECAP *********************************************************************
localhost                  : ok=11   changed=3    unreachable=0    failed=0

```

**but**

```
[vagrant@localhost ~]$ docker ps
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
```

FYI:
```
[vagrant@localhost ~]$ head host play.yml
==> host <==
[docker]
localhost ansible_connection=local

==> play.yml <==
- hosts: docker
  roles:
  - role: docker
    # Optional:
    docker_use_upstream_repo: True
```

**Comment by manics** [comment]
```
TASK [docker : docker | enable] ************************************************
changed: [localhost]
```
Should've enabled and started it. What does `systemctl status docker -l` and/or `journalctl --no-pager -l -u docker` show?

**Comment by joshmoore** [comment]
```
[vagrant@localhost ~]$ sudo systemctl status docker -l
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled)
   Active: active (running) since Mon 2015-11-23 10:34:11 EST; 9h ago
     Docs: https://docs.docker.com
 Main PID: 13821 (docker)
   CGroup: /system.slice/docker.service
           └─13821 /usr/bin/docker daemon -H fd://

Nov 23 10:34:07 localhost.localdomain docker[13821]: time="2015-11-23T10:34:07.198409914-05:00" level=info msg="API listen on /var/run/docker.sock"
Nov 23 10:34:07 localhost.localdomain docker[13821]: time="2015-11-23T10:34:07.435705145-05:00" level=warning msg="Usage of loopback devices is strongly discouraged for production use. Please use `--storage-opt dm.thinpooldev` or use `man docker` to refer to dm.thinpooldev section."
Nov 23 10:34:10 localhost.localdomain docker[13821]: time="2015-11-23T10:34:10.918424625-05:00" level=warning msg="Running modprobe bridge br_netfilter failed with message: modprobe: WARNING: Module br_netfilter not found.\ninsmod /lib/modules/3.10.0-229.14.1.el7.x86_64/kernel/net/llc/llc.ko \ninsmod /lib/modules/3.10.0-229.14.1.el7.x86_64/kernel/net/802/stp.ko \ninsmod /lib/modules/3.10.0-229.14.1.el7.x86_64/kernel/net/bridge/bridge.ko \n, error: exit status 1"
Nov 23 10:34:10 localhost.localdomain docker[13821]: time="2015-11-23T10:34:10.976413535-05:00" level=info msg="Firewalld running: false"
Nov 23 10:34:11 localhost.localdomain docker[13821]: time="2015-11-23T10:34:11.086042861-05:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Nov 23 10:34:11 localhost.localdomain docker[13821]: time="2015-11-23T10:34:11.256378870-05:00" level=info msg="Loading containers: start."
Nov 23 10:34:11 localhost.localdomain docker[13821]: time="2015-11-23T10:34:11.257295887-05:00" level=info msg="Loading containers: done."
Nov 23 10:34:11 localhost.localdomain docker[13821]: time="2015-11-23T10:34:11.257314628-05:00" level=info msg="Daemon has completed initialization"
Nov 23 10:34:11 localhost.localdomain docker[13821]: time="2015-11-23T10:34:11.257328968-05:00" level=info msg="Docker daemon" commit=a34a1d5 execdriver=native-0.2 graphdriver=devicemapper version=1.9.1
Nov 23 10:34:11 localhost.localdomain systemd[1]: Started Docker Application Container Engine.
```

```

**Comment by manics** [comment]
Did you add the user (`vagrant`?) to `docker_groupmembers`? If not doues `sudo docker` work?

**Comment by joshmoore** [comment]
Bingo. I assume this is something it's too complicated to work into the role itself?

**Comment by manics** [comment]
I can't think of an easy way to decide who should be in the group.

**Comment by joshmoore** [comment]
Running against minke with `-C` for check:

```
TASK [docker : upstream docker | setup repository] *****************************
changed: [minke.openmicroscopy.org]

TASK [docker : upstream docker | install docker] *******************************
fatal: [minke.openmicroscopy.org]: FAILED! => {"changed": false, "failed": true, "msg": "No Package matching 'docker-engine' found available, installed or updated", "rc": 0, "results": []}

PLAY RECAP *********************************************************************
minke.openmicroscopy.org   : ok=3    changed=1    unreachable=0    failed=1
```

Probably only a minor issue that will go away once it's been run without check, but something to keep in mind for future runs of `-C`.

**Comment by joshmoore** [comment]
First run against minke with `--ask-become-pass`:

```
TASK [docker : include] ********************************************************
included: /home/omero/infra/ansible/roles/docker/tasks/advanced-common.yml for minke.openmicroscopy.org

TASK [docker : storage | setup lvm docker-pool] ********************************
skipping: [minke.openmicroscopy.org]

TASK [docker : include] ********************************************************
skipping: [minke.openmicroscopy.org]

TASK [docker : include] ********************************************************
included: /home/omero/infra/ansible/roles/docker/tasks/advanced-upstream.yml for minke.openmicroscopy.org

TASK [docker : upstream docker | configure advanced] ***************************
skipping: [minke.openmicroscopy.org]

TASK [docker : Check change-time of original docker.service] *******************
ok: [minke.openmicroscopy.org]

TASK [docker : Check change-time of modified docker.service] *******************
ok: [minke.openmicroscopy.org]

TASK [docker : upstream docker | copy systemd file] ****************************
skipping: [minke.openmicroscopy.org]

TASK [docker : upstream docker | customise systemd file] ***********************
skipping: [minke.openmicroscopy.org]

TASK [docker : upstream docker | remove custom systemd file] *******************
ok: [minke.openmicroscopy.org]

TASK [docker : include] ********************************************************
included: /home/omero/infra/ansible/roles/docker/tasks/docker-common.yml for minke.openmicroscopy.org

TASK [docker : docker | enable] ************************************************
fatal: [minke.openmicroscopy.org]: FAILED! => {"changed": false, "failed": true, "msg": "Job for docker.service failed. See 'systemctl status docker.service' and 'journalctl -xn' for details.\n"}

PLAY RECAP *********************************************************************
minke.openmicroscopy.org   : ok=10   changed=2    unreachable=0    failed=1
```

**Comment by joshmoore** [comment]
and the error:

```
Nov 24 14:14:19 minke docker[48203]: /usr/bin/docker: relocation error: /usr/bin/docker: symbol dm_task_get_info_with_deferred_remove, version Base not defined...e reference
Nov 24 14:14:19 minke systemd[1]: docker.service: main process exited, code=exited, status=127/n/a
Nov 24 14:14:19 minke systemd[1]: Failed to start Docker Application Container Engine.
Nov 24 14:14:19 minke systemd[1]: Unit docker.service entered failed state.
```

**Comment by manics** [comment]
I think the docker rpm has a missing dependency on one of the device-mapper packages.
`  - role: upgrade-distpackages`?

**Comment by joshmoore** [comment]
```
TASK [docker : docker | enable] ************************************************
changed: [minke.openmicroscopy.org]

TASK [docker : docker | group members] *****************************************
skipping: [minke.openmicroscopy.org] => (item=docker_groupmembers)

PLAY RECAP *********************************************************************
minke.openmicroscopy.org   : ok=12   changed=2    unreachable=0    failed=0

[jamoore@minke ~]$ grep docker /etc/group
docker:x:502:jamoore
[jamoore@minke ~]$ sudo systemctl restart docker
[jamoore@minke ~]$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

:thumbsup:

**Comment by manics** [comment]
OK. Are you planning on using the advanced storage config? I'll help if you want.

**Comment by joshmoore** [comment]
What's appropriate to get this to production with the internal network?

**Comment by manics** [comment]
Basically all the storage should be on raw LVM volumes, and avoid using the loopback files. The network config can stay as default.

**Comment by joshmoore** [comment]
Ok. Merging this as tested. Can continue via gh-37.

**Pull request merged**